### DISCUSSION TITLE
Implicit `this` return for Builder Pattern

### DISCUSSION THREAD

**Discussion by JohnDesley** [dotnet/csharplang#7231] [Language Ideas] [answered]
# Implicit `this` return for Builder Pattern

## Summary
[summary]: #summary
This proposal suggests a language feature that allows methods to implicitly `return this` to enhance the builder pattern in C#. It introduces the keyword `this` as a return type, which indicates that the method automatically returns the current instance of the object (this). This would eliminate the need for explicit `return this` statements in builder methods, thereby making the code more readable and concise.

## Motivation
[motivation]: #motivation
The builder pattern is widely used in C# for creating complex objects step-by-step. However, the current implementation requires explicit `return this` statements at the end of each builder method to allow for method chaining. This redundancy can lead to more verbose code and potential mistakes if a developer forgets to include the return statement. By allowing an implicit return of `this`, we can improve the fluency of the builder pattern, making the code more readable and less error-prone.

## Detailed Design
[design]: #detailed-design
The proposed feature involves a new usage of `this` as a return type. When used in this way, `this` implies that the method will return the current instance of the object (this). This can be used in methods within any class, but is particularly useful in the builder pattern, where chained methods are common.

Here is an example of how the new feature could be used:
```csharp
public class Builder
{
    private int foo;

    // Using the proposed 'this' return type
    public this SetFoo(int bar) => foo = bar;

    ...
}
```

This code will be equivalent to:
```csharp
public class Builder
{
    private int foo;

    // The current way of implementing a builder method
    public Builder SetFoo(int bar) 
    {
        foo = bar;
        return this;
    }

    ...
}
```

By simply using `this` as the return type, developers can eliminate the need for `return this` at the end of each builder method, leading to cleaner and more readable code.

## Drawbacks
[drawbacks]: #drawbacks
One potential drawback of this proposal is that it introduces a new usage of the `this` keyword, which could lead to confusion for developers who are used to `this` being used to reference the current instance of an object. It could also complicate the language grammar and potentially conflict with other features.

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/902 https://github.com/dotnet/csharplang/issues/311 https://github.com/dotnet/csharplang/discussions/252

**Reply by TahirAhmadov** [reply]
Also, https://github.com/dotnet/csharplang/issues/5413

**Reply by sk0va** [reply]
Also #169
